File: javascript/sort/merge/merge.js
'use strict';

const merge = (array, p, q, r) => {
    const lowHalf = array.slice(p, q + 1);
    const highHalf = array.slice(q + 1, r + 1);

    let i = 0;
    let j = 0;
    let k = p;

    while (i < lowHalf.length && j < highHalf.length) {
        if (lowHalf[i] < highHalf[j]) {
            array[k] = lowHalf[i];
            i++;
        } else {
            array[k] = highHalf[j];
            j++;
        }

        k++;
    }

    while (i < lowHalf.length) {
        array[k] = lowHalf[i];
        i++;
        k++;
    }

    while (j < highHalf.length) {
        array[k] = highHalf[j];
        j++;
        k++;
    }
};

const mergeSort = (array, p, r) => {
    if (p >= r) {
        return array;
    }

    const q = Math.floor((p + r) / 2);

    mergeSort(array, p, q);
    mergeSort(array, q + 1, r);
    merge(array, p, q, r);

    return array;
};

const array = [14, 7, 3, 12, 9, 11, 6, 2];

console.log('array to sort:', array, '\n');
console.log('      results:', mergeSort(array, 0, array.length - 1));

console.assert(array[0] === 2);
console.assert(array[array.length - 1] === 14);

File: go/sort/merge/merge.go
package main

import "fmt"

var primes = []int{29, 19, 5, 17, 3, 11, 13, 2, 23, 7}

func merge(p, q, r int) {
	i, j, k := 0, 0, p
	lowHalf := make([]int, q-p+1) // Contains q-p+1 elements.
	highHalf := make([]int, r-q)  // Contains r-q elements.

	copy(lowHalf, primes[p:q+1])
	copy(highHalf, primes[q+1:r+1])

	for i < len(lowHalf) && j < len(highHalf) {
		if lowHalf[i] < highHalf[j] {
			primes[k] = lowHalf[i]
			i++
		} else {
			primes[k] = highHalf[j]
			j++
		}
		k++
	}

	for i < len(lowHalf) {
		primes[k] = lowHalf[i]
		i++
		k++
	}

	for j < len(highHalf) {
		primes[k] = highHalf[j]
		j++
		k++
	}
}

func mergeSort(p, r int) {
	if p < r {
		q := (p + r) / 2
		mergeSort(p, q)
		mergeSort(q+1, r)
		merge(p, q, r)
	}
}

func main() {
	fmt.Println(primes)
	mergeSort(0, len(primes)-1)
	fmt.Println(primes)
}
